subsys/emds/emds_flash.c

```diff
@@ -4,34 +4,428 @@
+
+#define ADDR_OFFS_MASK 0x0000FFFF
+#define EMDS_FLASH_BLOCK_SIZE 32
+
+/* Allocation Table Entry */
+struct shdn_ate {
+	uint16_t id; /* data id */
+	uint16_t offset; /* data offset within sector */
+	uint16_t len; /* data len within sector */
+	uint8_t crc8_data; /* crc8 check of the data entry */
+	uint8_t crc8; /* crc8 check of the ate entry */
+} __packed;
+
+enum ate_type {
+	ATE_TYPE_VALID = BIT(0),
+	ATE_TYPE_INVALIDATED = BIT(1),
+	ATE_TYPE_ERASED = BIT(2),
+	ATE_TYPE_UNKNOWN = BIT(3)
+};
+
+BUILD_ASSERT(offsetof(struct shdn_ate, crc8) == sizeof(struct shdn_ate) - sizeof(uint8_t),
+	     "crc8 must be the last member");
+
+static inline size_t align_size(struct emds_fs *fs, size_t len)
+{
+	uint8_t write_block_size = fs->flash_params->write_block_size;
+
+	if (write_block_size <= 1U) {
+		return len;
+	}
+	return (len + (write_block_size - 1U)) & ~(write_block_size - 1U);
+}
+
+static int ate_wrt(struct emds_fs *fs, const struct shdn_ate *entry)
+{
+	if (sizeof(struct shdn_ate) % fs->flash_params->write_block_size) {
+		return -EINVAL;
+	}
+
+	int rc = flash_write(fs->flash_dev, fs->ate_wra, entry, sizeof(struct shdn_ate));
+
+	fs->ate_wra -= fs->ate_size;
+	return rc;
+}
+
+static int data_wrt(struct emds_fs *fs, const void *data, size_t len)
+{
+	const uint8_t *data8 = (const uint8_t *)data;
+	int rc = 0;
+	off_t offset;
+	size_t blen;
+	size_t temp_len = len;
+	uint8_t buf[EMDS_FLASH_BLOCK_SIZE];
+
+	if (!temp_len) {
+		/* Nothing to write, avoid changing the flash protection */
+		return 0;
+	}
+
+	offset = fs->offset;
+	offset += fs->data_wra_offset & ADDR_OFFS_MASK;
+
+	blen = temp_len & ~(fs->flash_params->write_block_size - 1U);
+	/* Writes multiples of 4 bytes to flash */
+	if (blen > 0) {
+		rc = flash_write(fs->flash_dev, offset, data8, blen);
+		if (rc) {
+			return rc;
+		}
+
+		temp_len -= blen;
+		offset += blen;
+		data8 += blen;
+	}
+
+	if (temp_len) {
+		memcpy(buf, data8, temp_len);
+		(void)memset(buf + temp_len, fs->flash_params->erase_value,
+			     fs->flash_params->write_block_size - temp_len);
+		rc = flash_write(fs->flash_dev, offset, buf, fs->flash_params->write_block_size);
+		if (rc) {
+			return rc;
+		}
+	}
+
+	fs->data_wra_offset += align_size(fs, len);
+	return 0;
+}
+
+static int check_erased(struct emds_fs *fs, uint32_t addr, size_t len)
+{
+	size_t bytes_to_cmp;
+	uint8_t cmp[EMDS_FLASH_BLOCK_SIZE];
+	uint8_t buf[EMDS_FLASH_BLOCK_SIZE];
+
+	(void)memset(cmp, 0xff, EMDS_FLASH_BLOCK_SIZE);
+	while (len) {
+		bytes_to_cmp = MIN(EMDS_FLASH_BLOCK_SIZE, len);
+		if (flash_read(fs->flash_dev, addr, buf, bytes_to_cmp)) {
+			return -EIO;
+		}
+
+		if (memcmp(cmp, buf, bytes_to_cmp)) {
+			return 1;
+		}
+
+		len -= bytes_to_cmp;
+		addr += bytes_to_cmp;
+	}
+	return 0;
+}
+
+static int is_ate_valid(const struct shdn_ate *entry)
+{
+	return entry->crc8 == crc8_ccitt(0xff, entry, offsetof(struct shdn_ate, crc8));
+}
+
+static int entry_wrt(struct emds_fs *fs, uint16_t id, const void *data, size_t len)
```

Comment:
(Discussed in person) Well, it is a little tricky because of alignment to flash. To be able to write flash with a single write operation I belive the memory area must be a continious block. As far as I know it becomes delicate as soon as we want to append information to an incoming memory block going into flash, since the flash module does not have inate ownership over this memory area. I currently don't see a good solution to how we can enforce a user to provide memory blocks where a certain amount of space is reserved for block alignment, ATE entries, etc. The alternative is of course to allocate a  memory inside the flash module where we copy the data and do whatever we want with the rest, but this becomes a resource problem if the module needs to handle large chunks of data. 

It will only require three writes if the data block is not divisable by the block size of 4, so it is not an additional write for all cases. As we concluded we will accept that it is what it is at this point, complete the task, and then see what the performance is. The preliminary testing shows that the flash module will be able to store the worst case RPL list in about 13 ms, which is well within the limit of 30 ms we got from the customer.